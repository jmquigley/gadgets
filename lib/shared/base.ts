import * as React from 'react';
import {Size} from './index';

const styles = require('./styles.css');

export interface BaseOptions {
	className?: boolean;
	disabled?: boolean;
	visible?: boolean;
}

const defaultBaseOptions: BaseOptions = {
	className: true,
	disabled: true,
	visible: true
};

/**
 * A base class between React and all components in the module.
 */
export abstract class BaseComponent<P, S> extends React.Component<P, S> {

	protected _classes: string[] = [];
	protected _inlineStyle: any = {};    // inline style overrides
	protected _styles: any = {};         // css modules styles

	constructor(props: P, pstyles: any = {}) {
		super(props);
		this._styles = pstyles;
	}

	get classes(): string[] {
		return this._classes;
	}

	set classes(str: string[]) {
		this._classes = str;
	}

	get inlineStyle(): any {
		return this._inlineStyle;
	}

	set inlineStyle(val: any) {
		this._inlineStyle = Object.assign(this._inlineStyle, val);
	}

	get style(): string {
		return this._inlineStyle;
	}

	get styles(): any {
		return this._styles;
	}

	protected buildStyles(props: P, style: any = {}, opts?: BaseOptions): void {
		this._classes = [];
		this._inlineStyle = Object.assign(this._inlineStyle, props['style'], style);

		opts = Object.assign(
			defaultBaseOptions,
			opts
		);

		if (props['className'] !== '' && opts.className) {
			this._classes.push(props['className']);
		}

		if (!props['visible'] && opts.visible) {
			this._classes.push(this.styles.invisible);
		}

		if (props['disabled'] && opts.disabled) {
			this._classes.push(styles.disabled);
			this._classes.push('nohover');
		}
	}

	/**
	 * Takes a sizing parameter and a styles object and searches for the
	 * corresponding style class CSS.  If it is found, then it returns the local
	 * style for that size.
	 * @returns {string} the name of the local style for that size.  If
	 * it is not found, then an empty string is returned.
	 */
	protected getSizeStyle() {

		switch (this.props['size']) {
		case Size.xxsmall:
			if ('xxsmall' in this.styles) {
				return this.styles.xxsmall;
			}
			break;

		case Size.xsmall:
			if ('xsmall' in this.styles) {
				return this.styles.xsmall;
			}
			break;

		case Size.small:
			if ('small' in this.styles) {
				return this.styles.small;
			}
			break;

		case Size.large:
			if ('large' in this.styles) {
				return this.styles.large;
			}
			break;

		case Size.xlarge:
			if ('xlarge' in this.styles) {
				return this.styles.xlarge;
			}
			break;

		case Size.xxlarge:
			if ('xxlarge' in this.styles) {
				return this.styles.xxlarge;
			}
			break;

		case Size.normal:
		case Size.medium:
		default:
			if ('medium' in this.styles) {
				return this.styles.medium;
			}
		}

		return '';
	}
}
