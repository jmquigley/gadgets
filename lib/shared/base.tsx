/**
 * The base class for all components in the library.  This enables each module
 * to guarantee certain variables will be present through inheritance.
 *
 * The values of these variables are computed automatically for any component
 * that inherits from BaseComponent (controlled by props).  This class inherits
 * from `React.PureComponent` to take advantage of `shouldComponentUpdate`
 * shallow object comparison when computing styles based on props/state.
 *
 * #### Examples:
 *
 * ```javascript
 * import {BaseComponent} from '../shared';
 * ...
 * export class XYZ extends BaseComponent<Props, State> {
 *    ...
 * }
 *
 * ...
 *
 * <XYZ sizing={Sizing.xxsmall} location={Location.topRight} />
 * ```
 *
 * In the example above the `sizing` and `location` would be computed
 * automatically for the given values.  These values are then available to the
 * child class to use in building the component using these styles.
 *
 * @module BaseComponent
 */

const debug = require("debug");

import {isEqual} from "lodash";
import {Children, cloneElement, PureComponent} from "react";
import {DefaultTheme} from "styled-components";
import {calc} from "util.calc";
import {ClassNames} from "util.classnames";
import {getUUID} from "util.toolbox";
import {isDebug} from "./helpers";
import {KeyHandler, KeyMap} from "./keybinding";
import {BaseProps} from "./props";
import {FontStyle, Sizes, Sizing, Styling} from "./sizing";

require("./styles.css");

export const baseZIndex: number = 9999;
export const defaultSize: number = 16;
export let sizes: Sizes = Sizes.instance(defaultSize);

interface RenderOptions {
	noclassnameupdate?: boolean;
}

export abstract class BaseComponent<
	P extends BaseProps,
	S
> extends PureComponent<P, S> {
	private _className: ClassNames = new ClassNames();
	private _name: string;
	private _debug: any = null;
	private _debugCreate: any = null;
	private _debugRender: any = null;
	private _defaultClassName: string;
	public static defaultStyles: any = {};
	private _id: string;
	protected _keyHandler: KeyHandler = {};
	protected _keyMap: KeyMap = {};
	public state: S;

	constructor(defaultClassName: string, cls: any, props: P, state: S = null) {
		super(props);
		this.state = state || ({} as S);

		this._name = cls.name || "Unknown";
		BaseComponent.defaultStyles = cls.defaultProps.style || {};

		this._debug = debug(`gadgets.${this._name || "base"}`);
		this._debugCreate = debug(`gadgets.${this._name || "base"}:create`);
		this._debugRender = debug(`gadgets.${this._name || "base"}:render`);

		this._defaultClassName = defaultClassName;
		this._className = new ClassNames(defaultClassName);

		// If an id value is not given as a prop, then generate a unique id.  If the
		// component is under test, then 0 is used for the UUID value (to make it
		// predictable
		if (this.props.id) {
			this._id = this.props.id;
		} else {
			this._id =
				this._name + "-" + (this.props.testing ? "0" : getUUID());
		}

		if (isDebug()) {
			this._debugCreate("props: %O, state: %O", this.props, this.state);
		}
	}

	get className(): string {
		return this._className.value;
	}

	get defaultClassName(): string {
		return this._defaultClassName;
	}

	get defaultSize(): number {
		return defaultSize;
	}

	/**
	 * @return {string} the unique id value that was generated for this component
	 */
	get id(): string {
		return this._id;
	}

	get keyHandler(): KeyHandler {
		return this._keyHandler;
	}

	get keyMap(): KeyMap {
		return this._keyMap;
	}

	get name(): string {
		return this._name;
	}

	get theme(): DefaultTheme {
		return this.props.theme;
	}

	/**
	 * The input mappings structure is the name of the keyboaard mapping
	 * property and the associated function handler pointer.  These are
	 * mapped to their respective KeyMap and Handler objects.  These can
	 * then be used by the HotKeys object in a component to map a
	 * keyboard handler to that component.  The format of the input
	 * object is:
	 *
	 *     mappings = {
	 *         "kbActivate": this.handler
	 *     }
	 *
	 * The key name in the interface must match the prop name for that
	 * key binding.  See the <Button>
	 *
	 * @param mappings {KeyHanlder} - an object of key names and their
	 * associated function mappings.
	 */
	protected buildKeyMap(mappings: KeyHandler) {
		for (const key of Object.keys(mappings)) {
			this._keyMap[key] = this.props[key];
			this._keyHandler[key] = mappings[key];
		}
	}

	protected debug(...args: any[]) {
		if (isDebug()) {
			this._debug(...args);
		}
	}

	public static font(sizing: Sizing = Sizing.normal): FontStyle {
		return sizes[sizing].font;
	}

	public static fontSize(sizing: Sizing = Sizing.normal): number {
		return sizes[sizing].font.size;
	}

	public static fontSizePX(
		sizing: Sizing = Sizing.normal,
		scale: number = 1.0
	): string {
		return calc(sizes[sizing].font.sizepx, `* ${scale}`);
	}

	public static fontSizeREM(
		sizing: Sizing = Sizing.normal,
		scale: number = 1.0
	): string {
		return calc(sizes[sizing].font.sizerem, `* ${scale}`);
	}

	/**
	 * Each component has a getDerivedStateFromProps call.  This method is used
	 * by that call to set state properties that are common to all components.  It
	 * is checked for changes before the update will go through (so updates can be
	 * avoided when not needed).  The method contains a 3rd parameter flag to
	 * force this to return a new object instead of null.
	 * @param props {P} - the set of props that will be updated
	 * @param state {S} - the current state object when called
	 * @param forceUpdate=false {boolean} - force the state to return an object for
	 * update
	 * @return {S} a new, mutated state that will be merged into the current state
	 */
	public static getDerivedStateFromProps(
		props: any,
		state: any,
		forceUpdate: boolean = false
	): any {
		let newState: any = null;
		let dirty: boolean = forceUpdate;

		if (state && props) {
			newState = {...state};

			if ("style" in newState) {
				newState.style = {
					...BaseComponent.defaultStyles,
					...newState.style,
					...props.style
				};

				if (!isEqual(newState.style, state.style)) {
					dirty = true;
				}

				if (props.color) {
					newState.style["color"] = props.color;
					dirty = true;
				}

				if (props.backgroundColor) {
					newState.style["backgroundColor"] = props.backgroundColor;
					dirty = true;
				}

				if (props.borderColor) {
					newState.style["borderColor"] = props.borderColor;
					dirty = true;
				}
			}

			if ("sizing" in newState && props.sizing !== newState.sizing) {
				newState.sizing = props.sizing;
				dirty = true;
			}
		}

		if (dirty) {
			return newState;
		} else {
			return null;
		}
	}

	public static lineHeightPX(sizing: Sizing = Sizing.normal): string {
		return sizes[sizing].font.lineHeight;
	}

	/**
	 * Takes the given base Sizing and determines what the next size
	 * in the list would be.  e.g. if current size is *normal*, then
	 * the next size would be *large*.
	 * @param sizing {Sizing} an optional parameter that allows for overriding
	 * the default sizing when the class is created.
	 * @returns a reference to a `Styling` object.  This contains
	 * font type/size, box, and border information.
	 */
	public static next(sizing: Sizing = Sizing.normal): Styling {
		switch (sizing) {
			case Sizing.xxsmall:
				return sizes[Sizing.xsmall];
			case Sizing.xsmall:
				return sizes[Sizing.small];
			case Sizing.small:
				return sizes[Sizing.normal];
			case Sizing.large:
				return sizes[Sizing.xlarge];
			case Sizing.xlarge:
				return sizes[Sizing.xxlarge];
			case Sizing.xxlarge:
				return sizes[Sizing.xxlarge];

			case Sizing.normal:
			case Sizing.medium:
				return sizes[Sizing.large];

			default:
				return sizes[Sizing.normal];
		}
	}

	/**
	 * Takes the given base Sizing and determines what the previous size
	 * in the list would be.  e.g. if current size is *normal*, then
	 * the previous size would be *small*.
	 * @param sizing {Sizing} an optional parameter that allows for overriding
	 * the default sizing when the class is created.
	 * @returns a reference to a `Styling` object.  This contains font type/size,
	 * box, and border information.
	 */
	public static prev(sizing: Sizing = Sizing.normal): Styling {
		switch (sizing) {
			case Sizing.xxsmall:
				return sizes[Sizing.xxsmall];
			case Sizing.xsmall:
				return sizes[Sizing.xxsmall];
			case Sizing.small:
				return sizes[Sizing.xsmall];
			case Sizing.large:
				return sizes[Sizing.normal];
			case Sizing.xlarge:
				return sizes[Sizing.large];
			case Sizing.xxlarge:
				return sizes[Sizing.xlarge];

			case Sizing.normal:
			case Sizing.medium:
				return sizes[Sizing.small];

			default:
				return sizes[Sizing.normal];
		}
	}

	/**
	 * Called by the child render methods as the first call.  This is used to
	 * update base items before the child class performs its render operation.
	 */
	public render(options: RenderOptions = {}): any {
		options = {
			noclassnameupdate: false,
			...options
		};

		if (!options.noclassnameupdate) {
			this.updateClassName();
		}

		if (isDebug()) {
			this._debugRender("props: %O, state: %O", this.props, this.state);
		}
	}

	/**
	 * Checks the previous and incoming props for sizing changes.  If the control
	 * has children, then they are each updated with the sizing change.
	 * @param props {P} the previous properties on the component
	 * @param nextPrpos {P} the new properties for the component
	 * @return {any} a new set of children if the size has changed.  If the props
	 * do not have chilren, then NULL is returned.
	 */
	protected resizeChildren(props: P, nextProps: P): any {
		if (props.children && nextProps.children) {
			if (props.sizing !== nextProps.sizing) {
				// debug('Children prop sizing change');
				return Children.map(nextProps.children, (child: any) => {
					if (child.props && "sizing" in child.props) {
						return cloneElement(child, {
							sizing: nextProps.sizing
						});
					} else {
						return child;
					}
				});
			}
		}

		// debug('no children props to resize: props: %O, nextProps: %O', props, nextProps);
		return nextProps.children;
	}

	public static styling(sizing: Sizing = Sizing.normal): Styling {
		return sizes[sizing];
	}

	/**
	 * Returns the Sizing enum value associated with the given sizing.
	 * @param sizing {Sizing} an optional parameter that allows for overriding
	 * the default sizing when the class is created.
	 * @returns a refernce to a Sizing enum value.
	 */
	public static type(sizing: Sizing = Sizing.normal): Sizing {
		return sizes[sizing].type;
	}

	/**
	 * Used in each render() call for a component to properly setup up the
	 * className property used in that component
	 * @param classNames=null {string|string[]} - Allows the render method
	 * to add additional classes when building the default className
	 * @return a string that represents the current className string for
	 * a component.
	 */
	protected updateClassName(classNames: string | string[] = null): string {
		this._className.clear();
		this._className.add(this._defaultClassName);
		this._className.add(classNames);

		if (
			this.props.className &&
			!this._className.contains(this.props.className)
		) {
			this._className.add(this.props.className);
		}

		this._className.onIf(this.props.ripple && !this.props.disabled)(
			"ripple"
		);
		this._className.onIf(this.props.disabled)("nohover");

		if ("selected" in this.props) {
			this._className.onIf(this.props.selected)("ui-selected");
		}

		return this._className.value;
	}
}
