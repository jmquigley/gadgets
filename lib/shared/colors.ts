export type ColorType = string;

export enum ColorScheme {
	c1 = "#004358",
	c2 = "#1f8a70",
	c3 = "#bedb39",
	c4 = "#ffe11a",
	c5 = "#fd7400"
}

export enum Color {
	black = "#000000",
	error = "#d9534f",
	glow = "#51cbee",
	gray = "#666666",
	info = "#5bc0de",
	lightgray = "#cccccc",
	lightsaber = "#2ff923",
	red = "#ff0000",
	silver = "#c0c0c0",
	slategray = "#708090",
	success = "#5cb85c",
	ultralightgray = "#e6e6e6",
	warning = "#f0ad4e",
	white = "#ffffff"
}
