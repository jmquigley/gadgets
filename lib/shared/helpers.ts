"use strict";

const debug = require("debug")("helpers");

/**
 * Takes a variable name and an object and places that name and associated
 * object into the global node space.  If the object is already in the
 * global space, then that reference is used and returned.  The replace
 * flag is used to override the current global reference.
 * @param name {string} the name of the variable to store globally
 * @param pkg {object} the object associated with the variable name
 * @param replace {boolean} if true, then it will always overwrite the
 * current global (if it exists)
 * @return {object} the global instance reference
 */
export function globalize(
	name: string,
	pkgname: any,
	replace: boolean = false
) {
	let ref: any = pkgname;
	if (!(global as any)[name] || replace) {
		(window as any)[name] = (global as any)[name] = pkgname;
		debug("Adding global variable %o -> %O", name, (global as any)[name]);
	} else {
		ref = (global as any)[name];
	}

	return ref;
}
