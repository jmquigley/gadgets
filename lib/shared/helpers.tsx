'use strict';

const debugEntry = require('debug');
const pkg = require('../../package.json');

/**
 * A wrapper for the debug function.  It uses the package.json to enable or
 * disable debugging information in the library.  It also turns off
 * coverage checks for this function.
 * @param context {string} the name of the module where this debug request
 * is made.
 * @param args {object[]} the list of arguments passed to the the debug
 * function.
 */
export function debug(context: string, ...args: any[]) {
	/* istanbul ignore if  */
	if (pkg.debug) {
		const debugFn = debugEntry(`Gadgets -> ${context}`);
		debugFn(...args);
	}
}

/**
 * Takes a variable name and an object and places that name and associated
 * object into the global node space.  If the object is alread in the
 * global space, then that reference is used and returned.  The replace
 * flag is used to override the current global reference.
 * @param name {string} the name of the variable to store globally
 * @param pkg {object} the object associated with the variable name
 * @param replace {boolean} if true, then it will always overwrite the
 * current global (if it exists)
 * @return {object} the global instance reference
 */
export function globalize(name: string, pkgname: any, replace: boolean = false) {
	let ref: any = pkgname;
	if (!(global as any)[name] || replace) {
		(window as any)[name] = (global as any)[name] = pkgname;
	} else {
		ref = (global as any)[name];
	}

	return ref;
}
