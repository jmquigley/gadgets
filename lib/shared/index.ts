import {
	BaseComponent,
	baseZIndex,
	defaultSize
} from './base';

import {
	getDefaultWrapperProps,
	Wrapper,
	WrapperProps
} from './wrapper';

import {
	BaseProps,
	Color,
	ColorScheme,
	Direction,
	disabled,
	DisabledCSS,
	getDefaultBaseProps,
	invisible,
	InvisibleCSS,
	Justify,
	Location,
	locationStyle,
	SortOrder,
	Styles
} from './props';

import {
	borderStyle,
	boxStyle,
	fontStyle,
	FontStyle,
	lineHeightStyle,
	rectStyle,
	Sizes,
	Sizing,
	Styling
} from './sizing';

import {
	BaseState,
	getDefaultBaseState
} from './state';

import {
	getTheme,
	getThemeList,
	setTheme,
	Theme,
	ThemeProps
} from './themes';

export {
	BaseComponent,
	BaseProps,
	BaseState,
	baseZIndex,
	borderStyle,
	boxStyle,
	Color,
	ColorScheme,
	defaultSize,
	Direction,
	disabled,
	DisabledCSS,
	fontStyle,
	FontStyle,
	getDefaultBaseProps,
	getDefaultBaseState,
	getDefaultWrapperProps,
	getTheme,
	getThemeList,
	invisible,
	InvisibleCSS,
	Justify,
	lineHeightStyle,
	Location,
	locationStyle,
	rectStyle,
	setTheme,
	Sizes,
	Sizing,
	SortOrder,
	Styles,
	Styling,
	Theme,
	ThemeProps,
	Wrapper,
	WrapperProps
};
