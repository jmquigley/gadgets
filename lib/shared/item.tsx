'use strict';

import {cloneDeep} from 'lodash';
import * as React from 'react';
import {BaseProps, getDefaultBaseProps} from './props';
import {Title, TitleProps} from '../title';

const styles = require('./styles.css');

export interface ItemProps extends BaseProps, TitleProps {
	hiddenLeftButton?: boolean;
	hiddenRightButton?: boolean;
	leftButton?: any;
	rightButton?: any;
	selected?: boolean;
	title?: string;
}

export function getDefaultItemProps(): ItemProps {
	return cloneDeep(Object.assign(
		getDefaultBaseProps(), {
			hiddenLeftButton: false,
			hiddenRightButton: false,
			leftButton: null,
			rightButton: null,
			selected: false,
			stacked: false,
			title: '',
			widget: null
		}));
}

export const ItemComponent = (props: ItemProps) => {

	let leftButton = null;
	if (props.leftButton != null && !props.disabled) {
		leftButton = (
			<div className={`${styles.itemButton} ${(props.hiddenLeftButton) ? styles.hiddenButton : ''}`}>
				{props.leftButton != null ? props.leftButton : null}
			</div>
		);
	}

	let rightButton = null;
	if (props.rightButton != null && !props.disabled) {
		rightButton = (
			<div className={`${styles.itemButton} ${(props.hiddenRightButton) ? styles.hiddenButton : ''}`}>
				{props.rightButton != null ? props.rightButton : null}
			</div>
		);
	}

	return (
		<li
			id={props.id}
			onMouseOut={props.onMouseOut}
			onKeyDown={props.onKeyDown}
			onKeyPress={props.onKeyPress}
			className={`ui-item ${styles.item} ${props.selected ? 'ui-selected' : ''} ${props.classes.join(' ')}`}>
			{leftButton}
			<Title {...props} className={styles.itemTitle} classes={[]}>{props.title}</Title>
			{rightButton}
		</li>
	);
};
