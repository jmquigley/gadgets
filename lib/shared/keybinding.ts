/*
 * Interface defnintion used by the key map and bindings for components.  The
 * library uses the [react-hotkeys](https://www.npmjs.com/package/react-hotkeys)
 * module.  Note that all keybindings are on that component (not global), so
 * the components require focus to use the given binding.
 *
 * @module keybinding
 */

import {KeyMapOptions} from "react-hotkeys";

export interface KeyMap {
	[key: string]: string | KeyMapOptions;
}

export interface KeyHandler {
	[key: string]: (...args) => void;
}
