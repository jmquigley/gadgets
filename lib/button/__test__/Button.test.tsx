'use strict';

import {mount} from 'enzyme';
import * as React from 'react';
import * as sinon from 'sinon';
import {Sizing} from '../../shared';
import {Button, getDefaultButtonProps} from '../index';

test('Test retrieval of Button props object', () => {
	const props = getDefaultButtonProps();

	expect(props).toBeTruthy();

	expect('iconName' in props).toBe(true);
	expect(props.iconName).toBe('bomb');

	expect('iconStyle' in props).toBe(true);
	expect(props.iconStyle).toBe('');

	expect('sizing' in props).toBe(true);
	expect(props.sizing).toBe(Sizing.normal);
});

test('Test creation of a Button control', () => {
	const ctl = mount(<Button className="test-class"/>);

	expect(ctl).toBeTruthy();

	expect(ctl.prop('iconName')).toBe('bomb');
	expect(ctl.prop('disabled')).toBe(false);
	expect(ctl.prop('visible')).toBe(true);

	expect(ctl.find('.ui-button').length).toBe(1);
	expect(ctl.find('.fa').length).toBe(1);
	expect(ctl.find('.fa-bomb').length).toBe(1);
	expect(ctl.find('.test-class').length).toBe(1);
});

test('Test creation of a Button control with custom icon, colors, and border', () => {
	const ctl = mount(
		<Button
			color="red"
			backgroundColor="black"
			borderColor="green"
			borderWidth="2px"
			iconName="superpowers"
		/>
	);

	expect(ctl);

	expect(ctl.prop('iconName')).toBe('superpowers');
	expect(ctl.prop('color')).toBe('red');
	expect(ctl.prop('backgroundColor')).toBe('black');
	expect(ctl.prop('borderColor')).toBe('green');
	expect(ctl.prop('borderWidth')).toBe('2px');

	expect(ctl.prop('disabled')).toBe(false);
	expect(ctl.prop('visible')).toBe(true);

	expect(ctl.find('.ui-button').length).toBe(1);
	expect(ctl.find('.fa').length).toBe(1);
	expect(ctl.find('.fa-superpowers').length).toBe(1);
});

test('Test Button click event', () => {
	const click = sinon.spy();
	const ctl = mount(<Button onClick={click} />);

	expect(ctl).toBeTruthy();

	expect(ctl.prop('iconName')).toBe('bomb');
	expect(ctl.prop('disabled')).toBe(false);
	expect(ctl.prop('visible')).toBe(true);

	ctl.find('i').simulate('click');
	expect(click.calledOnce).toBe(true);
});

test('Test disabling of a Button', () => {
	const click = sinon.spy();
	const ctl = mount(<Button onClick={click} disabled={true} />);

	expect(ctl).toBeTruthy();

	expect(ctl.prop('iconName')).toBe('bomb');
	expect(ctl.prop('disabled')).toBe(true);
	expect(ctl.prop('visible')).toBe(true);

	ctl.find('i').simulate('click');
	expect(click.neverCalledWith()).toBe(true);
});

test('Test making a Button invisible', () => {
	const click = sinon.spy();
	const ctl = mount(<Button onClick={click} visible={false} />);

	expect(ctl).toBeTruthy();

	expect(ctl.prop('iconName')).toBe('bomb');
	expect(ctl.prop('disabled')).toBe(false);
	expect(ctl.prop('visible')).toBe(false);

	ctl.find('i').simulate('click');
	expect(click.neverCalledWith()).toBe(true);
});
