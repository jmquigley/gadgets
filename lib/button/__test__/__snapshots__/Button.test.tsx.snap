// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Button click event 1`] = `
<Button
  backgroundColor="inherit"
  borderColor="inherit"
  borderWidth="none"
  className=""
  color="inherit"
  contentEditable={false}
  direction="right"
  disabled={false}
  focus={false}
  iconName="bomb"
  iconStyle=""
  id=""
  location="none"
  noedit={false}
  nohover={false}
  noripple={false}
  onClick={[Function]}
  selected={false}
  sizing="normal"
  style={Object {}}
  visible={true}
>
  <div
    className="ui-button button ripple normal"
    onClick={[Function]}
    style={Object {}}
  >
    <Icon
      backgroundColor="inherit"
      borderColor="inherit"
      borderWidth="none"
      className=" icon"
      color="inherit"
      contentEditable={false}
      direction="right"
      disabled={false}
      focus={false}
      iconName="bomb"
      id=""
      imageFile=""
      location="none"
      noedit={false}
      nohover={false}
      noripple={false}
      selected={false}
      sizing="normal"
      style={Object {}}
      visible={true}
    >
      <i
        className="ui-icon icon none fa fa-bomb normal  icon"
        style={
          Object {
            "backgroundColor": "inherit",
            "color": "inherit",
          }
        }
      />
    </Icon>
  </div>
</Button>
`;

exports[`Test creation of a Button control 1`] = `
<div
  className="ui-button button ripple normal test-class"
  onClick={[Function]}
  style={Object {}}
>
  <Icon
    backgroundColor="inherit"
    borderColor="inherit"
    borderWidth="none"
    className=" icon"
    color="inherit"
    contentEditable={false}
    direction="right"
    disabled={false}
    focus={false}
    iconName="bomb"
    id=""
    imageFile=""
    location="none"
    noedit={false}
    nohover={false}
    noripple={false}
    selected={false}
    sizing="normal"
    style={Object {}}
    visible={true}
  />
</div>
`;

exports[`Test creation of a Button control with custom icon, colors, and border 1`] = `
<div
  className="ui-button button ripple normal"
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "black",
      "borderColor": "green",
      "borderWidth": "2px",
      "color": "red",
    }
  }
>
  <Icon
    backgroundColor="inherit"
    borderColor="inherit"
    borderWidth="none"
    className=" icon"
    color="inherit"
    contentEditable={false}
    direction="right"
    disabled={false}
    focus={false}
    iconName="superpowers"
    id=""
    imageFile=""
    location="none"
    noedit={false}
    nohover={false}
    noripple={false}
    selected={false}
    sizing="normal"
    style={Object {}}
    visible={true}
  />
</div>
`;

exports[`Test disabling of a Button 1`] = `
<Button
  backgroundColor="inherit"
  borderColor="inherit"
  borderWidth="none"
  className=""
  color="inherit"
  contentEditable={false}
  direction="right"
  disabled={true}
  focus={false}
  iconName="bomb"
  iconStyle=""
  id=""
  location="none"
  noedit={false}
  nohover={false}
  noripple={false}
  onClick={[Function]}
  selected={false}
  sizing="normal"
  style={Object {}}
  visible={true}
>
  <div
    className="ui-button button normal disabled nohover"
    onClick={[Function]}
    style={Object {}}
  >
    <Icon
      backgroundColor="inherit"
      borderColor="inherit"
      borderWidth="none"
      className=" icon"
      color="inherit"
      contentEditable={false}
      direction="right"
      disabled={false}
      focus={false}
      iconName="bomb"
      id=""
      imageFile=""
      location="none"
      noedit={false}
      nohover={false}
      noripple={false}
      selected={false}
      sizing="normal"
      style={Object {}}
      visible={true}
    >
      <i
        className="ui-icon icon none fa fa-bomb normal  icon"
        style={
          Object {
            "backgroundColor": "inherit",
            "color": "inherit",
          }
        }
      />
    </Icon>
  </div>
</Button>
`;

exports[`Test making a Button invisible 1`] = `
<Button
  backgroundColor="inherit"
  borderColor="inherit"
  borderWidth="none"
  className=""
  color="inherit"
  contentEditable={false}
  direction="right"
  disabled={false}
  focus={false}
  iconName="bomb"
  iconStyle=""
  id=""
  location="none"
  noedit={false}
  nohover={false}
  noripple={false}
  onClick={[Function]}
  selected={false}
  sizing="normal"
  style={Object {}}
  visible={false}
>
  <div
    className="ui-button button ripple normal invisible"
    onClick={[Function]}
    style={Object {}}
  >
    <Icon
      backgroundColor="inherit"
      borderColor="inherit"
      borderWidth="none"
      className=" icon"
      color="inherit"
      contentEditable={false}
      direction="right"
      disabled={false}
      focus={false}
      iconName="bomb"
      id=""
      imageFile=""
      location="none"
      noedit={false}
      nohover={false}
      noripple={false}
      selected={false}
      sizing="normal"
      style={Object {}}
      visible={true}
    >
      <i
        className="ui-icon icon none fa fa-bomb normal  icon"
        style={
          Object {
            "backgroundColor": "inherit",
            "color": "inherit",
          }
        }
      />
    </Icon>
  </div>
</Button>
`;

exports[`Test retrieval of Button props object 1`] = `
Object {
  "backgroundColor": "inherit",
  "borderColor": "inherit",
  "borderWidth": "none",
  "className": "",
  "color": "inherit",
  "contentEditable": false,
  "direction": "right",
  "disabled": false,
  "focus": false,
  "iconName": "bomb",
  "iconStyle": "",
  "id": "",
  "location": "none",
  "noedit": false,
  "nohover": false,
  "noripple": false,
  "onClick": [Function],
  "selected": false,
  "sizing": "normal",
  "style": Object {},
  "visible": true,
}
`;
